docs/configoption.rst

Configuration Options for peng3d
================================

Almost all important settings can be configured per-window or globally via the :py:data:`Peng.cfg` or :py:data:`Window.cfg` attributes.

Graphic Settings/OpenGL Base State
----------------------------------

For most of these graphical settings, it is important to actually use the exact type specified.
For example, you should only pass floats and not integers if the specified type is float.

.. confval:: graphics.clearColor
   
   A 4-tuple of RGBA colors used to clear the window before drawing.
   
   Each Color part should be a float between ``0`` and ``1``\ .
   
   By default, this option is set to ``(0.,0.,0.,1.)``\ .
   
   Be sure to verify that each value is a float, not an integer.

.. confval:: graphics.wireframe
   
   A Boolean value determining the polygon-fill-mode used by OpenGL.
   
   ``True`` results in ``GL_LINE`` being used, while ``False`` will result in ``GL_FILL`` being used.
   
   This option can be used to create a wireframe-like mode.
   
   The default value for this option is ``False``\ .
   
   .. note::
      
      This option is always turned off by :py:meth:`PengWindow.set2d()` but re-enabled by :py:meth:`PengWindow.set3d()` if necessary.

.. confval:: graphics.fieldofview
   
   An float value passed to :py:func:`gluPerspective()` as the first argument.
   
   For more information about this config option, see the GL/GLU documentation.
   
   By default, this option is set to ``65.0``\ .

.. confval:: graphics.nearclip
             graphics.farclip
   
   An float value specifying the near and far clipping plane, respectively.
   
   These clipping planes determine at what point vertices are cut off to save GPU cycles.
   
   By default, :confval:`graphics.nearclip` equals ``0.1`` and :confval:`graphics.farclip` equals ``10000``\ .

Fog settings
^^^^^^^^^^^^

.. confval:: graphics.fogSettings
   
   :py:class:`Config()` object storing the fog-specific settings.
   
   To access fog settings, use ``peng.cfg["graphics.fogSettings"]["<configoption>"]`` as appropriate.

.. confval:: graphics.fogSettings["enable"]
   
   A boolean value activating or deactivating the OpenGL fog.
   
   By default disabled.

.. confval:: graphics.fogSettings["color"]
   
   A 4-Tuple representing an RGB Color.
   
   Note that the values should be 0<=n<=1, not in range(0,256).
   
   For most cases, this value should be set to the clear color, else, visual artifacts may occur.

.. confval:: graphics.fogSettings["start"]
             graphics.fogSettings["end"]
   
   Defines start and end of the fog zone.
   
   The end value should be nearer than the far clipping plane to avoid cut-off vertices.
   
   Each value should be a float and is measured in standard OpenGL units.
   
   By default, the fog starts at 128 units and ends 32 units further out.

Light settings
^^^^^^^^^^^^^^

.. confval:: graphics.lightSettings
   
   :py:class:`Config()` object storing the light settings.
   
   To access light settings, use ``peng.cfg["graphics.lightSettings"]["<configoption>"]`` as appropriate.

.. confval:: graphics.lightSettings["enable"]
   
   A boolean value activating or deactivating the light config.
   
   By default disabled.

.. todo::
   
   Implement light settings with shader system

Controls
--------

Note that most of these config values are read when the appropriate objects are initialized,
this means that you should consult the objects documentation for how to change the option at runtime.

Mouse
^^^^^

.. confval:: controls.mouse.sensitivity
   
   Degrees to move per pixel traveled by the mouse.
   
   This applies to both horizontal and vertical movement.
   
   Defaults to ``0.15``\ .

Keyboard
^^^^^^^^

.. confval:: controls.controls.movespeed
   
   Speed multiplier for most movements.
   
   Defaults to ``10.0``\ .

.. confval:: controls.controls.verticalspeed
   
   Speed multiplier for vertical movement.
   
   Defaults to ``5.0``\ .

These keys are all registered with the ``mod`` flag set to False, thus they will ignore any modifiers.

.. confval:: controls.controls.forward
             controls.controls.backward
             controls.controls.strafeleft
             controls.controls.straferight
   
   Four basic movement keys.
   
   Each of these keys can be changed individually.
   
   Defaults are ``w``\ , ``s``\ , ``a`` and ``d``\, respectively.

.. confval:: controls.controls.jump
   
   Jump key.
   
   Defaults to ``space``\ .

.. confval:: controls.controls.crouch
   
   Crouch key.
   
   Defaults to ``lshift``\ .

Commonly used Key Combination Configuration Values
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. confval:: controls.keybinds.common.copy
             controls.keybinds.common.paste
             controls.keybinds.common.cut
   
   Key Combinations used to be used by various parts of the GUI.
   
   Currently used by the :py:class:`peng3d.gui.text.TextInput()` Widget for basic clipboard operations.
   
   By default, these are set to the commonly used values of ``Ctrl-C`` for Copy,
   ``Ctrl-V`` for Paste and ``Ctrl-X`` for Cutting.

General Controls Configuration Values
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. confval:: controls.keybinds.strict
   
   Whether or not keybindings should be strict.
   
   See :py:class:`peng3d.keybind.KeybindHandler()` for more information.

Debug Options
-------------

All of these options are disabled by default.

.. confval:: controls.keybinds.debug
   
   If enabled, all pressed keybinds will be printed.

.. confval:: debug.events.dump
   
   If enabled, all events are printed including their arguments.
   
   Note that ``on_draw`` and ``on_mouse_motion`` are never printed to avoid excessive outputs.

.. confval:: debug.events.logerr
   
   If enabled, Exceptions catched during event handling are printed.
   
   Note that only :py:exc:`AttributeError` exceptions are catched and printed, other exceptions will propagate further.

.. confval:: debug.events.register
   
   If enabled, all event handler registrations are printed.

.. confval:: debug.events.dumpfile
   
   If not an empty string, this should point to a valid file path for dumping all event names.
   
   If enabled, all event handler registrations and event sends will be logged to this file.
   Note that only the name of the event without data is stored and automatically deduplicated.
   
   Defaults to ``""``\ .

Resource Options
----------------

.. confval:: rsrc.enable
   
   Enables or Disables the resource module.
   
   By default enabled.

.. confval:: rsrc.basepath
   
   Base directory of the Resource Manager.
   
   By default determined via :py:func:`pyglet.resource.get_script_home()`\ .

.. confval:: rsrc.maxtexsize
   
   Maximum Texture size per bin.
   
   Limits the texture in size, useful if the graphics card has big textures (16kx16k) but only few textures will be needed.
   
   By default set to 1024.

.. _cfg-i18n:

Translation Options
-------------------

.. confval:: i18n.enable
   
   Enables or Disables the i18n module.
   
   By default enabled.

.. confval:: i18n.lang
   
   Determines the default language selected upon startup.
   
   Note that setting this config option after creating the first window will have
   no effect. Use :py:meth:`~peng3d.i18n.TranslationManager.setLang()` instead.
   
   Currently defaults to ``en``\ , but may be changed to operating system language
   in the future.

Event Options
-------------

.. confval:: events.removeonerror
   
   If True, automatically removes erroring event handlers.
   Note that the ``raiseErrors`` parameter takes precedent over this setting.
   
   Defaults to ``True``\ .

.. confval:: events.maxignore
   
   An integer number defining the maximum amount of ignored event messages to write to the log file.
   
   This setting is per event, not globally.
   
   Defaults to 3.

Other Options
-------------

.. confval:: pyglet.patch.patch_float2int
   
   Enables the float2int patch for pyglet.
   
   See :py:func:`peng3d.pyglet_patch.patch_float2int()` for more information.
   
   Enabled by default.

.. todo::
   
   Implement more config options


docs/api/peng3d.config.rst

``peng3d.config`` - Configuration system
========================================

.. automodule:: peng3d.config
   :members:
   :synopsis: Configuration system


peng3d/config.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
#  config.py
#  
#  Copyright 2016 notna <notna@apparat.org>
#  
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#  
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA 02110-1301, USA.
#  
#  

import sys
import os

__all__ = ["CFG_FOG_DEFAULT","CFG_LIGHT_DEFAULT","DEFAULT_CONFIG","Config"]

def _get_script_home():
    # Docs removed to not show on autodoc
    # Integrated from pyglet.resource to also work on headless distributions
    # see http://pyglet.readthedocs.io/en/pyglet-1.2-maintenance/programming_guide/resources.html
    frozen = getattr(sys, 'frozen', None)
    if frozen in ('windows_exe', 'console_exe'):
        return os.path.dirname(sys.executable)
    elif frozen == 'macosx_app':
        # py2app
        return os.environ['RESOURCEPATH']
    elif frozen:
        # pyinstaller
        return sys._MEIPASS
    else:
        main = sys.modules['__main__']
        if hasattr(main, '__file__'):
            return os.path.dirname(os.path.abspath(main.__file__))
        else:
            # cx_Freeze
            return os.path.dirname(sys.executable)

    # Probably interactive
    return os.getcwd() # Always return a path

class Config(object):
    """
    Configuration object imitating a dictionary.
    
    ``config`` can be any dictionary-style object and is used to store the configuration set by the user.
    This object only needs to implement the ``__getitem__``\ , ``__setitem__`` and ``__contains__`` special methods.
    
    ``defaults`` can be any dictionary-style object and is only read from in case the ``config`` object does not contain the key.
    Every config object is stackable, e.g. you can pass another :py:class:`Config` object as the ``defaults`` object.
    
    Example for stacking configs::
       
       >>> myconf = Config()
       >>> myconf2 = Config(defaults=myconf)
       >>> myconf["foo"] = "bar"
       >>> print(myconf2["foo"])
       bar
       >>> myconf2["bar"] = "foo"
       >>> print(myconf2["bar"])
       foo
       >>> print(myconf["bar"])
       Traceback (most recent call last):
       ...
       KeyError: Key "bar" does not exist
    
    There is no limit in stacking configurations, though higher-stacked configs may get slow when defaulting due to propagating through the whole chain.
    """
    
    def __init__(self,config=None,defaults={}):
        if config is None:
            config = {} # To avoid bugs with defaulted arguments being initialized at load time, not object instantiation.
        self.config = config
        self.defaults = defaults
    def __getitem__(self,key):
        return self.config[key] if key in self.config else self.defaults[key]
    def __setitem__(self,key,value):
        self.config[key] = value
    def __contains__(self,key):
        return key in self.config

CFG_FOG_DEFAULT = {
    "enable":False,
    
    "color":None,
    "start":128,
    "end":128+32,
    }
"""
Default fog configuration.

This configuration simply disables fog.
"""

CFG_LIGHT_DEFAULT = {"enable":False}
"""
Default lighting configuration.

This configuration simply disables lighting.
"""

DEFAULT_CONFIG = {
    # graphics.*
    # OpenGL config
    "graphics.clearColor":(0.,0.,0.,1.),
    "graphics.wireframe":False,
    "graphics.fieldofview":65.0,
    "graphics.nearclip":0.1,
    "graphics.farclip":10000, # It's over 9000!
    "graphics.min_size":None,
    "graphics.max_size":None,
    "graphics.stencil.enable":False,
    "graphics.stencil.bits":8,
    "graphics.fogSettings":Config({},defaults=CFG_FOG_DEFAULT),
    "graphics.lightSettings":Config({},defaults=CFG_LIGHT_DEFAULT),
    
    # controls.*
    # Controls
    "controls.mouse.sensitivity":0.15,
    
    "controls.controls.movespeed":10.0,
    "controls.controls.verticalspeed":5.0,
    "controls.controls.forward":"w",
    "controls.controls.backward":"s",
    "controls.controls.strafeleft":"a",
    "controls.controls.straferight":"d",
    "controls.controls.jump":"space",
    "controls.controls.crouch":"lshift",
    "controls.keybinds.strict":False,
    "controls.keybinds.debug":False,
    
    "controls.keybinds.common.copy":"Ctrl-C",
    "controls.keybinds.common.paste":"Ctrl-V",
    "controls.keybinds.common.cut":"Ctrl-X",
    
    # debug.*
    # Debug config
    "debug.events.dump":False,
    "debug.events.logerr":False,
    "debug.events.register":False,
    "debug.events.dumpfile":"",#"events.txt",
    
    # rsrc.*
    # Resource config
    "rsrc.enable":True,
    "rsrc.basepath":_get_script_home(),
    "rsrc.maxtexsize":1024, # Actual limit may be less, will be adjusted based on GL Capabilities
    
    # i18n.*
    # Translation config
    "i18n.enable":True,
    "i18n.lang":"en", # currently ISO 639-1 code, but may change in the future
    "i18n.lang.format":"{domain}:lang.{lang}", # format of resource that contains translations
    "i18n.lang.ext":".lang", # file extension including dot
    "i18n.discover_regex":r".*?/lang/(?P<lang>[a-zA-Z0-9_-]{2,5})\.lang",
    
    # events.*
    # Event config
    "events.removeonerror":True,
    "events.maxignore":3,
    
    # Other config options
    "pyglet.patch.patch_float2int":True,
}
"""
Default configuration values.

All default configuration values are stored here, for more information about specific config values, see :doc:`/configoption`\ .
"""
